supernova/src/ccs/mod.rs

```diff
@@ -0,0 +1,341 @@
+    /// `m` in the CCS/HyperNova papers.
+    pub num_constraints: usize,
+    /// Witness length.
+    ///
+    /// `m - l - 1` in the CCS/HyperNova papers.
+    pub num_vars: usize,
+    /// Length of the public input `X`. It is expected to have a leading
+    /// `ScalarField::ONE` element, thus this field must be non-zero.
+    ///
+    /// `l + 1`, w.r.t. the CCS/HyperNova papers.
+    pub num_io: usize,
+    /// Number of matrices.
+    ///
+    /// `t` in the CCS/HyperNova papers.
+    pub num_matrices: usize,
+    /// Number of multisets.
+    ///
+    /// `q` in the CCS/HyperNova papers.
+    pub num_multisets: usize,
+    /// Max cardinality of the multisets.
+    ///
+    /// `d` in the CCS/HyperNova papers.
+    pub max_cardinality: usize,
+    /// Set of constraint matrices.
+    pub Ms: Vec<SparseMatrix<G::ScalarField>>,
+    /// Multisets of selector indices, each paired with a constant multiplier.
+    pub cSs: Vec<(G::ScalarField, Vec<usize>)>,
+}
+
+impl<G: CurveGroup> CCSShape<G> {
+    /// Create an object of type `CCSShape` from the specified matrices and constant data structures
+    pub fn new(
+        num_constraints: usize,
+        num_vars: usize,
+        num_io: usize,
+        num_matrices: usize,
+        num_multisets: usize,
+        max_cardinality: usize,
+        Ms: Vec<SparseMatrix<G::ScalarField>>,
+        cSs: Vec<(G::ScalarField, Vec<usize>)>,
+    ) -> Result<CCSShape<G>, Error> {
+        if num_io == 0 {
+            return Err(Error::InvalidInputLength);
+        }
+
+        assert_eq!(Ms.len(), num_matrices);
+        assert_eq!(cSs.len(), num_multisets);
+
+        for (_c, S) in cSs.iter() {
+            if S.len() > max_cardinality {
+                return Err(Error::MultisetCardinalityMismatch);
+            }
+
+            S.iter().try_for_each(|idx| {
+                if idx >= &num_matrices {
+                    Err(Error::InvalidMultiset)
+                } else {
+                    Ok(())
+                }
+            })?;
+        }
+
+        Ok(Self {
+            num_constraints,
+            num_io,
+            num_vars,
+            num_matrices,
+            num_multisets,
+            max_cardinality,
+            Ms,
+            cSs,
+        })
+    }
+
+    /// Checks if the CCS instance together with the witness `W` satisfies the CCS constraints determined by `shape`.
+    pub fn is_satisfied<C: CommitmentScheme<G>>(
+        &self,
+        U: &CCSInstance<G, C>,
+        W: &CCSWitness<G>,
+        pp: &C::PP,
+    ) -> Result<(), Error> {
+        assert_eq!(W.W.len(), self.num_vars);
+        assert_eq!(U.X.len(), self.num_io);
+
+        let z = [U.X.as_slice(), W.W.as_slice()].concat();
+
+        let mut acc = vec![G::ScalarField::ZERO; self.num_constraints];
+        for (c, S) in &self.cSs {
+            let mut circle_product = vec![*c; self.num_constraints];
+
+            for idx in S {
+                let Mz = self.Ms[*idx].multiply_vec(&z);
```

Comment:
Plonkish constraints might have matrices appearing more than once (the standard Plonk constraint does, for example), so this would potentially save a bit of work. However, I don't think this will be a performance bottleneck, e.g. compared to the MSM.